pypesto/visualize/observable_mapping.py

```diff
@@ -37,10 +41,346 @@
+        for which to visualize the estimated observable mapping.
+    axes:
+        The axes to plot the estimated observable mapping on.
+    kwargs:
+        Additional arguments to passed to `matplotlib.pyplot.subplots`
+        (e.g. `figsize= ...`).
+
+    Returns
+    -------
+    axes:
+        The matplotlib axes.
+    """
+
+    # Check if the pyPESTO problem is hierarchical.
+    if not isinstance(pypesto_problem, HierarchicalProblem):
+        raise ValueError(
+            "Only hierarchical problems contain estimated observable mappings. Please provide a hierarchical problem."
+        )
+
+    amici_model = pypesto_problem.objective.amici_model
+
+    # Get the number of relative and semi-quantitative observables.
+    relative_obs_ids = pypesto_problem.relative_observable_ids or []
+    semiquant_obs_ids = pypesto_problem.semiquant_observable_ids or []
+    n_relative_observables = len(relative_obs_ids)
+    n_semiquant_observables = len(semiquant_obs_ids)
+
+    # Check if there are any relative or semi-quantitative observables.
+    if n_relative_observables == 0 and n_semiquant_observables == 0:
+        raise ValueError(
+            "The problem does not contain any relative or semi-quantitative observables."
+        )
+
+    # Get observable indices for both relative and semi-quantitative observables.
+    rel_and_semiquant_obs_indices = [
+        amici_model.getObservableIds().index(observable_id)
+        for observable_id in relative_obs_ids + semiquant_obs_ids
+    ]
+    rel_and_semiquant_obs_indices.sort()
+
+    # If axes are given, check if they are of the correct length.
+    if (
+        axes is not None
+        and len(axes) != n_relative_observables + n_semiquant_observables
+    ):
+        raise ValueError(
+            "The number of axes must be equal to the number of relative and semi-quantitative observables."
+        )
+
+    # If axes are not given, create them.
+    if axes is None:
+        n_axes = n_relative_observables + n_semiquant_observables
+        n_rows = int(np.ceil(np.sqrt(n_axes)))
+        n_cols = int(np.ceil(n_axes / n_rows))
+        _, axes = plt.subplots(n_rows, n_cols, **kwargs)
+        axes = axes.flatten()
+
+    # Plot the estimated observable mapping for relative observables.
+    if n_relative_observables > 0:
+        axes = plot_linear_observable_mappings_from_pypesto_result(
+            pypesto_result=pypesto_result,
+            pypesto_problem=pypesto_problem,
+            start_index=start_index,
+            axes=axes,
+            rel_and_semiquant_obs_indices=rel_and_semiquant_obs_indices,
+        )
+
+    # Plot the estimated spline approximations for semi-quantitative observables.
+    if n_semiquant_observables > 0:
+        axes = plot_splines_from_pypesto_result(
+            pypesto_result=pypesto_result,
+            start_index=start_index,
+            axes=axes,
+            rel_and_semiquant_obs_indices=rel_and_semiquant_obs_indices,
+        )
```

Comment:
The main routine `visualize_estimated_observable_mapping` calls the linear mapping and spline one if there are relative or semiquantitative observables, respectively. However, the two lower-level routines can also be called directly.

The same axes object is being passed around and plotted on.